lib/srv/server/azure_installer.go

```diff
@@ -0,0 +1,72 @@
+package server
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute/v3"
+	"github.com/aws/aws-sdk-go/aws"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/cloud/azure"
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+// AzureInstallerConfig represents configuration for an Azure install script
+// executor.
+type AzureInstallerConfig struct {
+	Emitter     apievents.Emitter
+	AccessPoint auth.DiscoveryAccessPoint
+}
+
+// AzureInstaller handles running commands that install Teleport on Azure
+// virtual machines.
+type AzureInstaller struct {
+	AzureInstallerConfig
+}
+
+// NewAzureInstaller returns a new Azure installer.
+func NewAzureInstaller(cfg AzureInstallerConfig) *AzureInstaller {
+	return &AzureInstaller{
+		AzureInstallerConfig: cfg,
+	}
+}
+
+// AzureRunRequest combines parameters for running commands on a set of Azure
+// virtual machines.
+type AzureRunRequest struct {
+	Client          azure.RunCommandClient
+	Instances       []*armcompute.VirtualMachine
+	Params          []string
+	Region          string
+	ResourceGroup   string
+	ScriptName      string
+	PublicProxyAddr string
+}
+
+// Run runs a command on a set of virtual machines and then blocks until the
+// commands have completed.
+func (ai *AzureInstaller) Run(ctx context.Context, req AzureRunRequest) error {
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(10)
+
+	for _, inst := range req.Instances {
+		inst := inst
+		g.Go(func() error {
+			runRequest := azure.RunCommandRequest{
+				Region:        req.Region,
+				ResourceGroup: req.ResourceGroup,
+				VMName:        aws.StringValue(inst.Name),
+				Parameters:    req.Params,
+				Script:        getInstallerScript(req.ScriptName, req.PublicProxyAddr),
+			}
+			return trace.Wrap(req.Client.Run(ctx, runRequest))
+		})
+	}
+	return trace.Wrap(g.Wait())
+}
+
+func getInstallerScript(installerName, publicProxyAddr string) string {
+	return fmt.Sprintf("curl -s -L https://%s/v1/webapi/scripts/installer/%v | bash -s $@", publicProxyAddr, installerName)
```

Comment:
Is there a way to verify if we don't execute some random code? SHA-1 check?

Reply:
It would be a little messy to get all of the info needed to construct the script for hashing on the discovery service, but it's possible (although at that point we might as well send the script itself instead of the `curl` trick). As far as I know, EC2 auto-join does not verify the script.

Reply:
This makes me uneasy as well; the fact that EC2 doesn't verify anything makes me uneasy about it too.

Let's raise an issue and assign someone to follow it up. Surely we can do better than a blind curl to somewhere.